### ISSUE TITLE
Add SSL Certificate checks for General TCP ports

### ISSUE THREAD

**Issue by throwabird** [louislam/uptime-kuma#1079]
### ⚠️ Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find similar feature request

### 🏷️ Feature Request Type

New Notification, New Monitor

### 🔖 Feature description

Say you have an application running on a TCP port using an SSL Certificate that will expire eventually.
You want to have notifications for the certificate expiry (as is possible with HTTPS currently).

## Usecases
IMAP on Port 993
```sh
openssl s_client -connect imap.gmail.com:993
```
SMTP starttls checks on Port 25
```sh
openssl s_client -starttls smtp smtp.gmail.com:25
```

### ✔️ Solution

I propose Uptime Kuma adds (optional) SSL certificate monitoring for the `TCP Port` Monitor which might use similar functionality as described above.

### 📝 Additional Context

Trying to add expiry notifications for my mail server certificate.

**Comment by alexlehm** [comment]
I would like that as well, I am monitoring my gemini server which is ssl/1965 port and it would be nice to check the connect and maybe the cert validity since the server usually used a custom cert where only validity is checked

**Comment by woj-tek** [comment]
Yes, that would be great. Apart from IMAP mentioned already, XMPP also operates over TCP/SSL so it would help a lot.

There is also STARTTLS, but it's protocol dependent and I think there is general trend toward avoiding STARTTLS and just using DirectTLS (TCP/SSL)

**Comment by woj-tek** [comment]
I agree that that DirectTLS is easier (simply open encrypted socket) and STARTTLS would require handling for each protocol (the list available here: https://en.wikipedia.org/wiki/Opportunistic_TLS#SSL_ports). From my perspective I'm mostly interested in XMPP (I agree it's not the most popular protocols nowadays) and unfortunately STARTTLS is the most popular mechanism in it (DirectTLS is getting more popular but it's not there yet)

**Comment by alexlehm** [comment]
It should be easy to write a plugin for that for the monitor (though i have not looked at how plugins work in general)

**Comment by CommanderStorm** [comment]
> It should be easy to write a plugin for that for the monitor (though i have not looked at how plugins work in general, it might be worthwhile to write a few)

There are multiple proposals in issues. Until now there has been #1787 accepted, but more could follow.
This is a feature in the [ideas section of the Roadmap](https://github.com/users/louislam/projects/4/views/1).
I don't quite know what Louis vision is on this topic, but one plugin is available at https://uptime.kuma.pet/c/plugins.json

**Comment by sarfaraz-ecosmob** [comment]
I would like that as well, I am monitoring ssl on custom tcp port

**Comment by h4knet** [comment]
+1 Would be a nice feature to have !

**Comment by CommanderStorm** [comment]
@h4knet 

Please refrain from posting `+1` / requests for updates things on issues, as this makes issue-management harder.
Issues are for discussing **what needs to be done how by whom**.
[We use 👍🏻 on issues to prioritise work](https://github.com/louislam/uptime-kuma/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc), as always: [Pull Requests](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md) welcome.

**Comment by CommanderStorm** [comment]
@Techssh as mentioned above:
 
> Please refrain from posting `+1` / requests for updates things on issues, as this makes issue-management harder. Issues are for discussing **what needs to be done how by whom**. [We use 👍🏻 on issues to prioritise work](https://github.com/louislam/uptime-kuma/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc), as always: [Pull Requests](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md) welcome.

We would be open to such contributions.
See our [contribution guide](https://github.com/louislam/uptime-kuma/blob/master/CONTRIBUTING.md) and [`dns.js`](https://github.com/louislam/uptime-kuma/blob/master/server/monitor-types/dns.js)

**Comment by kolbma** [comment]
When implementing this, can you please add an option to compare public key of certificate with a stored public key in the monitor configuration.  
E.g. via sha256sum of the public key block.

A public key compare is the only possibility to check, that the cert has been requested by the private key owner and not by a 3d party with a different private key and acting as MITM.

**Comment by alexlehm** [comment]
> When implementing this, can you please add an option to compare public key of certificate with a stored public key in the monitor configuration. E.g. via sha256sum of the public key block.

That would be useful for both HTTPS checks and generic TLS checks

**Comment by CommanderStorm** [comment]
@kolbma
You can add your own CA to the list of trusted CAs via [`NODE_EXTRA_CA_CERTS=file`](https://nodejs.org/api/cli.html#node_extra_ca_certsfile).
What you are requesting is a different feature

**Comment by emerysteele** [comment]
Would really love this feature. My web/email server seems to have some quirks sometimes where the TLS cert will auto renew but will only be applied to the web server, and not the imap or smtp servers.

If TLS cert could be monitored for other ports, would be great help to catch it before it expires.

**Comment by windware-ono** [comment]
This feature is very useful in a case like the one mentioned in the above comment where there are many tools that can monitor validity of https certs but when it comes to mail servers, I'll probably have to deploy my own shell script to not miss on expirations not to cause email clients (especially automated ones) borking one day without notice.

**Comment by CommanderStorm** [comment]
> [!IMPORTANT]
> The main Usecases noted above are resolved by:
> - https://github.com/louislam/uptime-kuma/pull/5489

**Comment by dancojocaru2000** [comment]
How would one monitor the expiration of TLS certificates when it's not an HTTPS or SMTP service? In particular, I'd want to monitor the certificate of an XMPP server.

**Comment by tigunia-martin** [comment]
I'll +1 this.
I want to monitor the SSL cert of a FTPS server.
